src/mds/MDCache.cc

```diff
@@ -6900,8 +6900,11 @@ bool MDCache::trim_inode(CDentry *dn, CInode *in, CDir *con, map<mds_rank_t, MCa
     // This is because that unconnected replicas are problematic for
     // subtree migration.
     //
-    if (!in->is_auth() && !in->dirfragtreelock.can_read(-1))
+    if (!in->is_auth() &&
+        !(in->dirfragtreelock.can_read(-1) ||
+          mds->locker->rdlock_try(in->dirfragtreelock, -1, nullptr))) {
```

Comment:
can_read(-1) check is no needed any more, because rdlock_try also call calls can_read(-1)